feat(3DViewportPresets): 3D Volume Viewport Presets and Cinematic Rendering

IbrahimCSAE opened the pull request:
### Context

This adds 3D volume presets for the 3D volume viewport. It's accessible in the window-level action menu.

### Changes & Results

- VolumeLighting Component
- VolumeMappingRange Component
- VolumeQuality Component
- VolumeRenderingOptions Component
- VolumeRenderingPresets Component
- VolumeRenderingPresetsContent Component
- VolumeShade Component
- Draggable Modal
- Labels for input range
- Customization service for viewport presets
- Customization service for viewport quality
- setViewportPreset command
- setVolumeQuality command
- setVolumeMappingRange command
- setVolumeShade command
- setVolumeAmbient coomand
- setVolumeDiffuse command
- setVolumeSpecular command
- support for PNG in Icon component

### Screenshots

![image](https://github.com/OHIF/Viewers/assets/93064150/152a382a-8193-4e9d-bda1-aae8795f5a4d)

![image](https://github.com/OHIF/Viewers/assets/93064150/24aaa7a6-7411-407f-8c30-2ebebe918a2c)

![image](https://github.com/OHIF/Viewers/assets/93064150/23a30791-b0e0-4645-aba6-6b624fb9754c)


### Test here

http://localhost:3000/viewer?StudyInstanceUIDs=2.16.840.1.114362.1.11972228.22789312658.616067305.306.2&hangingProtocolId=only3D

sedghi commented:
Noice let me review

sedghi commented on platform/ui/src/assets/styles/styles.css:
```diff
@@ -1,7 +1,9 @@
 /* CUSTOM OHIF SCROLLBAR */
 .ohif-scrollbar {
-  scrollbar-color: #173239 transparent;
+  scrollbar-color: #041c4a transparent;
```
are we changing the scrollbar color all together in the app?

IbrahimCSAE replied:
I don't think the current scrollbar color fits with anything at the moment, that's why I decided to update it all together, if not a good idea then I can make it specific to the new changes.

sedghi commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -658,6 +658,18 @@ function commandsModule({
+      if (!viewport) {
+        return;
+      }
+      console.log(viewport, preset);
```
remove

IbrahimCSAE replied:
got it :+1:

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/WindowLevelActionMenu.tsx:
```diff
@@ -137,6 +150,21 @@ export function WindowLevelActionMenu({
             />
           </AllInOneMenu.SubMenu>
         )}
+
+        {viewportPresets && is3DVolume && (
+          <AllInOneMenu.SubMenu
+            key="viewportPresets"
+            itemLabel="Volume Rendering"
+            itemIcon="VolumeRendering"
+          >
+            <VolumeRendering
+              serviceManager={serviceManager}
+              viewportId={viewportId}
+              commandsManager={commandsManager}
+              viewportPresets={viewportPresets}
+            />
+          </AllInOneMenu.SubMenu>
+        )}
```
I'm not sure in case of 3D we need to show the colorLUT or the w/l presets

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeRendering.tsx:
```diff
@@ -0,0 +1,58 @@
+    });
+  };
+
+  const onClickCinematic = () => {
+    uiModalService.show({
+      content: CinematicRendering,
+      title: 'Cinematic Rendering',
+      movable: true,
+      contentProps: {
+        onClose: uiModalService.hide,
+        viewportId,
+        serviceManager,
+      },
+      customContainerDimensions: 'h-[180px] w-[460px]',
+      customContentDimensions: 'h-[130px] w-[460px]  pl-[12px]',
+    });
+  };
```
This should be part of the all in one menu not a separate pop up

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumePresets.tsx:
```diff
@@ -0,0 +1,106 @@
+    }
+  }, [searchValue, presets]);
+
+  const handleApply = useCallback(
+    props => {
+      commandsManager.runCommand('setViewportPreset', {
+        ...props,
+      });
+    },
+    [commandsManager]
+  );
```
can you move it to be close to the other usecallback

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumePresets.tsx:
```diff
@@ -0,0 +1,106 @@
+    return label.length > maxChars ? `${label.slice(0, maxChars)}...` : label;
+  };
+
+  useEffect(() => {
+    if (searchValue) {
+      const filtered = presets.filter(preset =>
+        preset.name.toLowerCase().includes(searchValue.toLowerCase())
+      );
+      setFilteredPresets(filtered);
+    } else {
+      setFilteredPresets(presets);
+    }
+  }, [searchValue, presets]);
```
I think this filtering can be inside the handleSearchChange callback, and we don't need a useeffect here

sedghi reviewed (commented):
please see my comments, I will try it after you added the rendering logics

IbrahimCSAE commented:
Volume Quality, Shift, and Width are complete

Pending is Shade/Ambient/Diffuse/Specular

sedghi commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -658,6 +658,54 @@ function commandsModule({
       });
     },
+
+    setViewportPreset: ({ viewportId, preset }) => {
```
please add documentation here

sedghi commented on extensions/cornerstone/src/getCustomizationModule.ts:
```diff
@@ -3,8 +3,15 @@ import { toolNames } from './initCornerstoneTools';
 
 const DefaultColormap = 'Grayscale';
+const { VIEWPORT_PRESETS } = CONSTANTS;
+const defaultVolumeQualityRange = {
```
we don't need to name this object, can you just use those in the bottom and not put it in the top?

sedghi commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -658,6 +658,116 @@ function commandsModule({
+      const { actor } = viewport.getActors()[0];
+      actor.getProperty().setShade(shade);
+      viewport.render();
+    },
+
+    /**
+     * Sets the ambient lighting level for volume rendering in a given viewport.
+     * @param {string} viewportId - The ID of the viewport to set the ambient lighting.
+     * @param {number} ambient - The ambient lighting level for the volume rendering.
+     */
+
+    setVolumeAmbient: ({ viewportId, ambient }) => {
+      const viewport = cornerstoneViewportService.getCornerstoneViewport(viewportId);
+      const { actor } = viewport.getActors()[0];
+      actor.getProperty().setAmbient(ambient);
+      viewport.render();
+    },
+
+    /**
+     * Sets the diffuse lighting level for volume rendering in a given viewport.
+     * @param {string} viewportId - The ID of the viewport to set the diffuse lighting.
+     * @param {number} diffuse - The diffuse lighting level for the volume rendering.
+     */
+    setVolumeDiffuse: ({ viewportId, diffuse }) => {
+      const viewport = cornerstoneViewportService.getCornerstoneViewport(viewportId);
+      const { actor } = viewport.getActors()[0];
+      actor.getProperty().setDiffuse(diffuse);
+      viewport.render();
+    },
+    /**
+     * Sets the specular lighting level for volume rendering in a given viewport.
+     * @param {string} viewportId - The ID of the viewport to set the specular lighting.
+     * @param {number} specular - The specular lighting level for the volume rendering.
+     */
+    setVolumeSpecular: ({ viewportId, specular }) => {
+      const viewport = cornerstoneViewportService.getCornerstoneViewport(viewportId);
+      const { actor } = viewport.getActors()[0];
+      actor.getProperty().setSpecular(specular);
+      viewport.render();
+    },
```
```
setVolumeLighting: ({ viewportId, options }) => {
  const viewport = cornerstoneViewportService.getCornerstoneViewport(viewportId);
  const { actor } = viewport.getActors()[0];
  const property = actor.getProperty();
  
  if (options.shade !== undefined) {
    property.setShade(options.shade);
  }
  
  if (options.ambient !== undefined) {
    property.setAmbient(options.ambient);
  }
  
  if (options.diffuse !== undefined) {
    property.setDiffuse(options.diffuse);
  }
  
  if (options.specular !== undefined) {
    property.setSpecular(options.specular);
  }
  
  viewport.render();
}
```

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeShade.tsx:
```diff
@@ -0,0 +1,39 @@
+    const { actor } = viewport.getActors()[0];
+    const shade = actor.getProperty().getShade();
+    setShade(shade);
+    setKey(key + 1);
```
this is a bit odd, is it used for re-rendeing?

IbrahimCSAE replied:
yes, to update the switch button properly.

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeQuality.tsx:
```diff
@@ -0,0 +1,69 @@
+import React, { ReactElement, useCallback, useState, useEffect } from 'react';
+import { VolumeQualityProps } from '../../types/ViewportPresets';
+
+export function VolumeQuality({
```
volumeRenderingQuality?

sedghi commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeQuality.tsx:
```diff
@@ -0,0 +1,69 @@
+
+  const onChange = useCallback(
+    (value: number) => {
+      commandsManager.runCommand('setVolumeQuality', {
```
setVolumeRenderingQulaity?

sedghi reviewed (commented):
is this the new design by @dan-rukas 

The gap seems too big really bw the icon and scrollbar and in the bottom it doesn't go to the end, please sync with Dan

![CleanShot 2024-03-12 at 09 10 43](https://github.com/OHIF/Viewers/assets/7490180/14088f6c-eeba-4713-8134-809ac6b8377a)

Icon actually looks too much to the right edge

![CleanShot 2024-03-12 at 09 15 07](https://github.com/OHIF/Viewers/assets/7490180/01f0aae4-9d1d-46d4-8801-15057329b385)


---

Also I thought we have moved to putting icons to the top then the label overlays don't need a gap in the bottom 

----

Also still colorbar design doesn't look correct, it it too thick can you ask Dan for per viewport width design for this?

![CleanShot 2024-03-12 at 09 12 26](https://github.com/OHIF/Viewers/assets/7490180/4013d0ef-c5d7-40f3-a830-7c4a427f61bb)

-----

The logo looks odd to me, it seems pixelated? is it PNG? it should be svg

![CleanShot 2024-03-12 at 09 13 23](https://github.com/OHIF/Viewers/assets/7490180/7c6d9fa7-2fe9-4fe2-8d8a-72193a964494)


---

The window level icon for inverted (pt) seems incorrect please ask Dan

![CleanShot 2024-03-12 at 09 13 58](https://github.com/OHIF/Viewers/assets/7490180/08d5110a-f4c6-4e47-9b8a-5e3b3c127fc8)


----

Fusion color bar is wrong in terms of positioniong, should be center

![CleanShot 2024-03-12 at 09 14 36](https://github.com/OHIF/Viewers/assets/7490180/0f458e5c-8627-4738-986f-be3b5d5c29ea)

---

This is a wrong behaviour for some reason, go to the following link, I can't really see the 3d presets

http://localhost:3000/viewer?StudyInstanceUIDs=2.16.840.1.114362.1.11972228.22789312658.616067305.306.2&hangingprotocolid=mprAnd3DVolumeViewport

I'm not able to see the 3d presets menu for some reason, I think we need to make sure that the above link works (without the layout changes), after you made it work i will try and comment the rest of the code 

Great work so far

IbrahimCSAE commented:
I requested designs from dan for:

- window level icon positioning
- gap removal at the bottom of the viewport
- scrollbar position
- colorbar position and style in single and dual mode
- window level icon for inverted pt

The OHIF icon is svg, so I'm not sure, this the icon I got from Dan.

I added some fixes and you can try it out now, I also addressed the comments 

http://localhost:3000/viewer?StudyInstanceUIDs=2.16.840.1.114362.1.11972228.22789312658.616067305.306.2&hangingprotocolid=mprAnd3DVolumeViewport

sedghi commented:
This icons doesn't look right, they seem smaller than expected

![CleanShot 2024-03-13 at 08 57 49](https://github.com/OHIF/Viewers/assets/7490180/9a681eeb-0577-4450-9837-ca9bfd2261e6)

---

Also when i load the menu for 3d preset, seems like some times it takes time to actually render the icon, can we use preloaders or something to make sure they load quickly enough? (not important, but the solution might be easy if you search)

https://github.com/OHIF/Viewers/assets/7490180/3289cc7f-45b3-44e6-b79b-4806c1d0ed22

----

I don't think the click outside rendering presets should close it, ask Dan, but I expect the user to want to rotate the view before closing either by X or Cancel, ask Dan

---

Still feels like the text is too thin, ask Dan to evaluate if a higher weight looks better 

![CleanShot 2024-03-13 at 09 01 25](https://github.com/OHIF/Viewers/assets/7490180/5b04f4df-a62d-4546-8ae4-3376442dde16)

---

There is this unusuall move if you drag a thumbnail accidently

https://github.com/OHIF/Viewers/assets/7490180/905330d2-8880-45e5-a5eb-29ebc17048ca

---


Shift and width are not doing as expected, see below for a correct implementation in 3D slicer, maybe we just put shift slider similar to 3D slicer


https://share.cleanshot.com/GmD1d0fN

IbrahimCSAE commented:
I updated design, reimplemented shift, and removed width.

IbrahimCSAE commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -658,6 +658,107 @@ function commandsModule({
+     * @param {string} viewportId - The ID of the viewport to set the mapping range.
+     * @param {number} shift - The shift value to shift the points by.
+     */
+    shiftVolumeOpacityPoints: ({ viewportId, shift }) => {
```
New Shift

IbrahimCSAE commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeRenderingOptions.tsx:
```diff
@@ -0,0 +1,46 @@
+        viewportId={viewportId}
+        commandsManager={commandsManager}
+        serviceManager={serviceManager}
+      />
```
Separator for lighting in design

![image](https://github.com/OHIF/Viewers/assets/93064150/2008fc95-dfa6-4a3e-9c74-3f9c616a6cf2)

IbrahimCSAE commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeShift.tsx:
```diff
@@ -0,0 +1,91 @@
+import React, { ReactElement, useCallback, useEffect, useState, useRef } from 'react';
+import { VolumeShiftProps } from '../../types/ViewportPresets';
+
+export function VolumeShift({
```
New shift

IbrahimCSAE commented on extensions/cornerstone/src/getCustomizationModule.ts:
```diff
@@ -47,12 +49,21 @@ function getCustomizationModule() {
         {
           id: 'cornerstone.colorbar',
-          width: '2.5%',
+          width: '16px',
```
Adjusted colorbar width from dan

IbrahimCSAE commented on extensions/cornerstone/src/services/ColorbarService/ColorbarService.ts:
```diff
@@ -228,7 +228,11 @@ export default class ColorbarService extends PubSubService {
   private createContainers(numContainers, element, position, thickness, viewportId) {
     const containers = [];
-    const dimension = 50 / numContainers;
+    const dimensions = {
```
Centered properly if there's two colorbars

IbrahimCSAE commented on platform/ui/src/components/AllInOneMenu/SubMenu.tsx:
```diff
@@ -21,22 +21,18 @@ const SubMenu = (props: SubMenuProps) => {
           name={props.itemIcon}
-          width="20px"
-          height="20px"
+          width="25px"
```
Bigger icons as requested

![image](https://github.com/OHIF/Viewers/assets/93064150/cf6f2bd9-a4b7-42e4-8fb2-beb98ac06f8d)

IbrahimCSAE commented on platform/ui/src/components/Modal/Modal.tsx:
```diff
@@ -20,58 +21,90 @@ const Modal = ({
-      </section>
+      {movable ? (
+        <Draggable
+          handle=".drag-handle"
```
Assigned a draggable handled to the modal, only the handle is draggable

IbrahimCSAE commented on platform/ui/src/components/ViewportActionCorners/ViewportActionCorners.tsx:
```diff
@@ -20,19 +20,19 @@ const commonClasses = 'pointer-events-auto flex items-center gap-1';
 const classes = {
   [ViewportActionCornersLocations.topLeft]: classNames(
```
fix action icons positing, and remove bottom spacing entirely as instructed by Dan

IbrahimCSAE commented:
@sedghi if there's design adjustments to be made again, no problem but lets merge this into the actual new-window-level branch and I'll continue there, I'm doing general changes for colorbar and other stuff that shouldn't be here.

sedghi reviewed (commented):
I don't understand why we need a two step (click and apply) to change a preset? why we can't just apply on the thumbnail select (see  below)

https://github.com/OHIF/Viewers/assets/7490180/691f052c-2182-400a-9e25-d8e24482ff34


Also can we make it so that the viewport is interactable while the modal is open? so that the user can rotate? (see above)

https://github.com/OHIF/Viewers/assets/7490180/cc6a36e8-ad20-42a0-bdb5-422a8bcefc95

Also seems like the shift range is basically very small area and the rest is kind of useless, can we just make the range smaller for shift then?

IbrahimCSAE commented on platform/ui/src/components/Modal/Modal.tsx:
```diff
@@ -20,58 +21,90 @@ const Modal = ({
+      }
+      overlayClassName={
+        movable
+          ? 'fixed top-0 left-0 right-0 bottom-0 z-50 flex items-center justify-center py-16 pointer-events-none'
```
pointer events disabled on modal overlay

IbrahimCSAE commented on platform/ui/src/components/Modal/Modal.tsx:
```diff
@@ -20,58 +21,90 @@ const Modal = ({
+      {movable ? (
+        <Draggable
+          handle=".drag-handle"
+          defaultClassName="bg-primary-dark pointer-events-auto"
```
point events enabled only on the modal itself

IbrahimCSAE commented on extensions/cornerstone/src/components/WindowLevelActionMenu/VolumeRenderingPresetsContent.tsx:
```diff
@@ -0,0 +1,95 @@
+              <div
+                key={index}
+                className="flex cursor-pointer flex-col items-start"
+                onClick={() => {
```
apply preset on thumbnail click

sedghi reviewed (approved):
Looks great, thanks
